opentracing-impl-java8/src/main/java/io/opentracing/propagation/AbstractTracer.java

```diff
@@ -0,0 +1,175 @@
+        return registry.getExtractor((Class<T>)carrier.getClass()).joinTrace(carrier);
+    }
+
+    public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+        return registry.register(carrierType, injector);
+    }
+
+    public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+        return registry.register(carrierType, extractor);
+    }
+
+    private static class PropagationRegistry {
+
+        private final ConcurrentMap<Class, Injector> injectors = new ConcurrentHashMap<>();
+        private final ConcurrentMap<Class, Extractor> extractors = new ConcurrentHashMap<>();
+
+        public <T> Injector<T> getInjector(Class<T> carrierType) {
+            Class<?> c = carrierType;
+            // match first on concrete classes
+            do {
+                if (injectors.containsKey(c)) {
+                    return injectors.get(c);
+                }
+                c = c.getSuperclass();
+            } while (c != null);
+            // match second on interfaces
+            for (Class<?> iface : carrierType.getInterfaces()) {
+                if (injectors.containsKey(c)) {
+                    return injectors.get(c);
+                }
+            }
+            throw new AssertionError("no registered injector for " + carrierType.getName());
+        }
+
+        public <T> Extractor<T> getExtractor(Class<T> carrierType) {
+            Class<?> c = carrierType;
+            // match first on concrete classes
+            do {
+                if (extractors.containsKey(c)) {
+                    return extractors.get(c);
+                }
+                c = c.getSuperclass();
+            } while (c != null);
+            // match second on interfaces
+            for (Class<?> iface : carrierType.getInterfaces()) {
+                if (extractors.containsKey(c)) {
+                    return extractors.get(c);
+                }
+            }
+            throw new AssertionError("no registered extractor for " + carrierType.getName());
+        }
+
+        public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+            return injectors.putIfAbsent(carrierType, injector);
+        }
+
+        public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+            return extractors.putIfAbsent(carrierType, extractor);
+        }
+    }
+
+    static class SpanBuilderImpl implements SpanBuilder {
+
+        SpanBuilderImpl() {}
+
+        final Map<String,String> traceState = new HashMap<>();
+        final Map<String,String> traceAttributes = new HashMap<>();
+
+        private String operationName = null;
+        private Span parent = null;
+        private Instant start = Instant.now();
+        private final Map<String,String> stringTags = new HashMap<>();
+        private final Map<String,Boolean> booleanTags = new HashMap<>();
+        private final Map<String,Number> numberTags = new HashMap<>();
```

Comment:
Simplifies the way `span.withTags(..)` is subsequently called.

I suspect this code will get redone after some performance profiling. But my initial look through AbstractMap's constructor and maybe it won't. Again I'll be happy to come back to it.